internal/controllers/scheduling/op.go

```diff
@@ -0,0 +1,233 @@
+package scheduling
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"hash/fnv"
+	"reflect"
+	"time"
+
+	apiv1 "github.com/Azure/eno/api/v1"
+	"github.com/go-logr/logr"
+	"github.com/google/uuid"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type op struct {
+	Synthesizer *apiv1.Synthesizer
+	Composition *apiv1.Composition
+	Reason      opReason
+
+	Dispatched time.Time
+
+	id               uuid.UUID // set when patch is built
+	synthRolloutHash []byte    // memoized
+}
+
+func newOp(synth *apiv1.Synthesizer, comp *apiv1.Composition) *op {
```

Comment:
This implements the logic of the old `shouldSwapStates` function - just (IMO) much easier to grok.

I promise it passes the old unit tests. They were wired up during refactoring but I removed them since the new tests provide better coverage anyway.